misc-bench/histogram-alt/hist-alt.cpp

```diff
@@ -0,0 +1,284 @@
+    assert(status == PIM_OK);
+    redCount[i] += prevCount;
+
+    std::cout << i << std::endl;
+  }
+  pimFree(imgObj);
+  pimFree(tempObj);
+}
+
+
+int main(int argc, char *argv[])
+{
+  struct Params params = getInputParams(argc, argv);
+  std::string fn = params.inputFile;
+  std::cout << "Input file : '" << fn << "'" << std::endl;
+  int fd;
+  uint64_t imgDataBytes;
+  int imgDataOffsetPosition;
+  struct stat finfo;
+  char* fdata;
+  unsigned short* dataPos;
+
+  // Start data parsing
+  if (fn.substr(fn.find_last_of(".")) != ".bmp")
+  {
+    // TODO: reading in other types of input files
+    std::cout << "Need work reading in other file types" << std::endl;
+    return 1;
+  }
+  else
+  {
+    // Assuming input will be 24-bit .bmp files
+    fd = open(params.inputFile.c_str(), O_RDONLY);
+    if (fd < 0) 
+    {
+      perror("Failed to open input file, or file doesn't exist");
+      return 1;
+    }
+    if (fstat(fd, &finfo) < 0) 
+    {
+      perror("Failed to get file info");
+      return 1;
+    }
+    fdata = static_cast<char *>(mmap(0, finfo.st_size + 1, PROT_READ, MAP_PRIVATE, fd, 0));
+    if (fdata == 0) 
+    {
+      perror("Failed to memory map the file");
+      return 1;
+    }
+
+    imgDataOffsetPosition = 10; // Start of image data, ignoring unneeded header data and info
+    // Defined according to the assumed input file structure given
+
+    dataPos = (unsigned short *)(&(fdata[imgDataOffsetPosition]));
+    imgDataBytes = static_cast<uint64_t> (finfo.st_size) - static_cast<uint64_t>(*(dataPos));
+  }
+  // End data parsing
+
+  printf("This file has %lu bytes of image data, %lu pixels\n", imgDataBytes, imgDataBytes / NUMCHANNELS);
+
+  if (!createDevice(params.configFile))
+  {
+    return 1;
+  }
+
+  PimDeviceProperties deviceProp;
+  PimStatus status = pimGetDeviceProperties(&deviceProp);
+  assert(status == PIM_OK);
+
+  std::vector<uint64_t> redCount(NUMBINS, 0), greenCount(NUMBINS, 0), blueCount(NUMBINS, 0);
+  std::vector<uint8_t> imgData(fdata + *dataPos, fdata + finfo.st_size);
+  std::vector<uint8_t> orderedImgData;
+
+  for (int i = 0; i < NUMCHANNELS; ++i) 
+  {
+    for (uint64_t j = 0; j < imgDataBytes; j+=NUMCHANNELS)
+    {
+      orderedImgData.push_back(imgData[i + j]);
+    }
+  }
+
+  uint64_t numCol = deviceProp.numColPerSubarray, numRow = deviceProp.numRowPerSubarray, 
+           numCore = deviceProp.numRanks * deviceProp.numBankPerRank * deviceProp.numSubarrayPerBank;
+  uint64_t totalAvailableBits = numCol * numRow * numCore;
+  uint64_t requiredBitsforImage = ((imgDataBytes / NUMCHANNELS * 8) + 8); // Using uint8_t instead of int, only require 8 bits
+  int numItr = std::ceil(static_cast<double> (requiredBitsforImage) / totalAvailableBits);
+  // std::cout << "Required iterations for image: " << numItr << std::endl;
+
+  if (numItr == 1)
+  {
+    histogram(imgDataBytes, orderedImgData, redCount, greenCount, blueCount);
+  }
+  else
```

Comment:
I think we can remove the conditional `if-else` and based on the `numItr`, we can run it in multiple pass